In [110]:
import numpy as np 
true_values_test = [5. , 10. , 15. , 5. , 17.]
predectid_values_test = [5., 9.,16.,2.,15. ]

**Implementing Mean Square Error (MSE)**

In [111]:
def mse (true_values , predectid_values):
  total_error = 0
  for yt , yp in zip(true_values, predectid_values):
    total_error += pow(yt-yp,2)
    
  mse =   total_error/len(true_values)
  print("MSE:" , mse)
  return mse

In [112]:
mse(true_values_test,predectid_values_test)

MSE: 3.0


3.0

In [113]:
mse = np.mean(pow(np.array(true_values_test)-np.array(predectid_values_test),2))
print(mse)

3.0


**Implementing Mean Absolute Error (MSE)**

In [114]:

def mae (true_values , predectid_values):
  total_error = 0
  for yt , yp in zip(true_values, predectid_values):
    total_error += abs(yt-yp)
    
  mae =   total_error/len(true_values)
  print("MAE:" , mae)
  return mae

In [115]:

mae(true_values_test,predectid_values_test)
     

MAE: 1.4


1.4

In [116]:
mae =np.mean(np.abs(np.array(true_values_test)-np.array(predectid_values_test)))
print(mae)

1.4


**Implementing Sigmoid Function**

In [98]:
def sigmoid(x):
   return 1/(1+np.exp(-x))



**Implementing Simple NN Project** 

In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from tensorflow import keras

In [68]:
df = pd.read_csv(r"C:\Users\dell\Desktop\summer_third_year\Deep_Learning_Projects\Neural-Network-From-Scratch\insurance_data.csv")

In [69]:
df.head()

age  affordibility  bought_insurance
0   22              1                 0
1   25              0                 0
2   47              1                 1
3   52              0                 0
4   46              1                 1

In [70]:
df.isnull().sum()

age                 0
affordibility       0
bought_insurance    0
dtype: int64

In [71]:
print(df.columns)

Index(['age', 'affordibility', 'bought_insurance'], dtype='object')


In [72]:
scaler = MinMaxScaler(feature_range=(0, 1))  # Apply scaling between 0 and 1 values 
df["age"] = scaler.fit_transform(df[["age"]])

In [73]:
df.head()

age  affordibility  bought_insurance
0  0.090909              1                 0
1  0.159091              0                 0
2  0.659091              1                 1
3  0.772727              0                 0
4  0.636364              1                 1

In [74]:
x_train , x_test , y_train , y_test = train_test_split(df[['age','affordibility']],df["bought_insurance"],random_state=42 , test_size=0.2 ,)

In [86]:
model = keras.Sequential([
  keras.layers.Dense(1 , input_shape=(2,) , activation="sigmoid" , kernel_initializer="ones" , bias_initializer="zeros"),

])


model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])
model.fit(x_train, y_train,epochs=5000)

Epoch 1/5000


c:\Users\dell\AppData\Local\Programs\Python\Python310\lib\site-packages\keras\src\layers\core\dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 417ms/step - accuracy: 0.5000 - loss: 0.7174
Epoch 2/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 54ms/step - accuracy: 0.5000 - loss: 0.7170
Epoch 3/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 51ms/step - accuracy: 0.5000 - loss: 0.7165
Epoch 4/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 32ms/step - accuracy: 0.5000 - loss: 0.7161
Epoch 5/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 34ms/step - accuracy: 0.5000 - loss: 0.7157
Epoch 6/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 52ms/step - accuracy: 0.5000 - loss: 0.7152
Epoch 7/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step - accuracy: 0.5000 - loss: 0.7148
Epoch 8/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 32ms/step - accuracy: 0.5000 - loss: 0.7144
Epoch 9/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - accuracy: 0.5000 - loss: 0.7139
Epoch 10/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step - accuracy: 0.5000 - loss: 0.7135
Epoch 11/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - accuracy: 0.5000 - loss: 0.7131
Epoch 12/5000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - accuracy: 0.5000 - lo

In [83]:
model.evaluate(x_test, y_test)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - accuracy: 1.0000 - loss: 0.1313


[0.13127177953720093, 1.0]

In [90]:
x_test , y_test

(         age  affordibility
 9   0.977273              1
 25  0.818182              1
 8   1.000000              1
 21  0.181818              0
 0   0.090909              1
 12  0.204545              0,
 9     1
 25    1
 8     1
 21    0
 0     0
 12    0
 Name: bought_insurance, dtype: int64)

In [89]:
model.predict(x_test)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 52ms/step


array([[0.9219695 ],
       [0.8589483 ],
       [0.9285152 ],
       [0.1007221 ],
       [0.22732149],
       [0.10962896]], dtype=float32)

Getting the values of weights and bias 
- coef -> Weights
- intercept -> bias

In [93]:
coef, intercept = model.get_weights()
print(coef, intercept)

[[4.16636  ]
 [1.3444891]] [-2.9467468]


Then 
- the weight of age (W1) = 4.16636 
- the weight of affordibility (W2) = 1.3444891 
- bias = -2.9467468

**We can now try our optamized weights and bias with a very simple prediction system**

In [94]:
def prediction_function(age, affordibility):
    weighted_sum = coef[0]*age + coef[1]*affordibility + intercept
    return sigmoid(weighted_sum)

print(prediction_function(0.977273, 1))

0.9219695657286794


C:\Users\dell\AppData\Local\Temp\ipykernel_984\3643866665.py:3: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  return 1 / (1 + math.exp(-x))


**Impleminting Gradient Descent from scratch** 

This function implements the **log loss (binary cross-entropy)**.  
It is used to evaluate the performance of a classification model where the output is a probability between 0 and 1.  


In [ ]:
# log loss function
def log_loss(y_true, y_predicted): 
    epsilon = 1e-15
    y_predicted_new = [max(i,epsilon) for i in y_predicted]
    y_predicted_new = [min(i,1-epsilon) for i in y_predicted_new]
    y_predicted_new = np.array(y_predicted_new)
    return -np.mean(y_true*np.log(y_predicted_new)+(1-y_true)*np.log(1-y_predicted_new))

This function implements **gradient descent** to optimize weights for a simple logistic regression model.

In [107]:
def gradient_descent(age, affordability, y_true, epochs, loss_thresold):
    w1 = w2 = 1
    bias = 0
    rate = 0.5
    n = len(age)
    for i in range(epochs):
        weighted_sum = w1 * age + w2 * affordability + bias
        y_predicted = sigmoid(weighted_sum)
        loss = log_loss(y_true, y_predicted)

        w1d = (1/n)*np.dot(np.transpose(age),(y_predicted-y_true)) 
        w2d = (1/n)*np.dot(np.transpose(affordability),(y_predicted-y_true)) 

        bias_d = np.mean(y_predicted-y_true)
        w1 = w1 - rate * w1d
        w2 = w2 - rate * w2d
        bias = bias - rate * bias_d

        print (f'Epoch:{i}, w1:{w1}, w2:{w2}, bias:{bias}, loss:{loss}')

        if loss<=loss_thresold:
            break

    return print("W1:",w1 ,"\nW2:", w2 ,"\nbias:", bias)

In [108]:
gradient_descent(x_train['age'],x_train['affordibility'],y_train,1000, 0.4402)


Epoch:0, w1:0.9797428980185209, w2:0.9274298940280011, bias:-0.12375264627590753, loss:0.7173831932516557
Epoch:1, w1:0.966712705681736, w2:0.8659918812059308, bias:-0.23143707438289124, loss:0.6782942243194876
Epoch:2, w1:0.9604240345326028, w2:0.8150822469167088, bias:-0.32428683069132236, loss:0.6495163989664631
Epoch:3, w1:0.9601731977564524, w2:0.7737118491510965, bias:-0.4039260727354506, loss:0.6287456015872123
Epoch:4, w1:0.9651478458458423, w2:0.7406886013547752, bias:-0.4721416975532016, loss:0.6138857105695502
Epoch:5, w1:0.9745201491152293, w2:0.7147736518360271, bias:-0.5307014651712642, loss:0.6032158895295766
Epoch:6, w1:0.9875105874895461, w2:0.6947887599769155, bias:-0.5812369676378858, loss:0.5954263310765129
Epoch:7, w1:1.0034229953982785, w2:0.6796759415277157, bias:-0.6251849346720276, loss:0.5895738429433302
Epoch:8, w1:1.021657891694067, w2:0.6685212106595032, bias:-0.6637702160534745, loss:0.5850059821862388
Epoch:9, w1:1.0417120249762348, w2:0.6605556148174913,

In [102]:
print(coef, intercept)

[[4.16636  ]
 [1.3444891]] [-2.9467468]
